proof-systems/src/darlin/poly-commit/src/constraints/data_structures.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{BDFGMultiPointProof, PCKey, PCVerifierState, PolynomialLabel};
+use algebra::{Group, PrimeField};
+use r1cs_std::groups::GroupGadget;
+use r1cs_std::prelude::AllocGadget;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+/// A commitment gadget plus its label, needed for reference.
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct LabeledCommitmentGadget<
+    ConstraintF: PrimeField,
+    G: Group<BaseField = ConstraintF>,
+    GG: GroupGadget<G, ConstraintF>,
+> {
+    label: PolynomialLabel,
+    commitment: GG,
+    _constraintf: PhantomData<ConstraintF>,
+    _group: PhantomData<G>,
+}
+
+impl<ConstraintF, G, GG> LabeledCommitmentGadget<ConstraintF, G, GG>
+where
+    ConstraintF: PrimeField,
+    G: Group<BaseField = ConstraintF>,
+    GG: GroupGadget<G, ConstraintF>,
+{
+    /// Instantiate a new labeled commitment from a label and a commitment gadget.
+    pub fn new(label: PolynomialLabel, commitment: GG) -> Self {
+        Self {
+            label,
+            commitment,
+            _constraintf: PhantomData::<ConstraintF>,
+            _group: PhantomData::<G>,
+        }
+    }
+
+    /// Return the label for `self`.
+    pub fn label(&self) -> &String {
+        &self.label
+    }
+
+    /// Retrieve the commitment from `self`.
+    pub fn commitment(&self) -> &GG {
+        &self.commitment
+    }
+}
+
+/// Define interface for a gadget representing an opening proof for a multi-point assertion
+pub trait MultiPointProofGadget<
+    ConstraintF: PrimeField,
+    G: Group<BaseField = ConstraintF>,
+    MPP: BDFGMultiPointProof<G>,
+>: AllocGadget<MPP, ConstraintF>
+{
+    /// Type of commitment gadget
+    type CommitmentGadget;
+    /// Type of the opening proof gadget for a single-point assertion
+    type ProofGadget;
+
+    /// get the proof gadget for the combined single-point assertion
+    fn get_proof(&self) -> &Self::ProofGadget;
+
+    /// get the commitment of polynomial h, which is computed in the opening proof of multi-point assertion
+    fn get_h_commitment(&self) -> &Self::CommitmentGadget;
+}
+
+/// Gadget for the state returned by verifier in case of successful verification
+pub trait VerifierStateGadget<VS: PCVerifierState, ConstraintF: PrimeField>:
+    Clone + Debug + Eq + PartialEq + AllocGadget<VS, ConstraintF>
+{
+}
+/// Interface for the gadget representing the verifier key
+pub trait VerifierKeyGadget<VK: PCKey, ConstraintF: PrimeField>:
+    Clone + Debug + Eq + PartialEq + AllocGadget<VK, ConstraintF>
```

Comment:
In theory there shouldn't be any need to allocate a Dlog key, as it should be the same throughout the scheme, so we can as well (and it would be more correct to) hardcode it @UlrichHaboeck75 .
This will allow to save:

- Checks on h and s when allocating vk;
- Computation of rand_times_s, in IPAPCGadget::succinct_verify() can be a fixed base mul;
- Computation of h_prime in IPAPCGadget::succinct_verify() can be a fixed base endo_mul (if it does make sense at all @UlrichHaboeck75 ).

Reply:
Ok, I removed the gadget and I updated the succinct_verify implementation for IPA both in the primitive and in the gadget.